pymedphys/_gui/streamlit/pseudonymise.py

```diff
@@ -0,0 +1,297 @@
+# limitations under the License.
+
+
+import base64
+import datetime
+import io
+import os
+import pathlib
+from os.path import join as pjoin
+from typing import List
+from zipfile import ZIP_DEFLATED, ZipFile
+
+import streamlit as st
+
+import pydicom
+
+from pymedphys._dicom.anonymise import anonymise_dataset
+from pymedphys._dicom.constants.core import DICOM_SOP_CLASS_NAMES_MODE_PREFIXES
+from pymedphys._dicom.utilities import remove_file
+from pymedphys.experimental import pseudonymisation as pseudonymisation_api
+
+# hack below copied from content in streamlit forum
+# HACK This only works when we've installed streamlit with pipenv, so the
+# permissions during install are the same as the running process
+STREAMLIT_STATIC_PATH = pathlib.Path(st.__path__[0]) / "static"
+if not STREAMLIT_STATIC_PATH.is_dir():
+    STREAMLIT_STATIC_PATH.mkdir()
+# We create a downloads directory within the streamlit static asset directory
+# and we write output files to it
+DOWNLOADS_PATH = STREAMLIT_STATIC_PATH / "downloads"
+if not DOWNLOADS_PATH.is_dir():
+    DOWNLOADS_PATH.mkdir()
+
+DELAY_TIME = 1
+
+
+def link_to_zipbuffer_download(filename: str, zip_bytes: bytes):
+    """Generates a link allowing the bytes to be downloaded using the specified filename
+    Assumes that the data is in zip format
+
+    Parameters
+    ----------
+    filename : str
+        default filename to use
+    zip_bytes : bytes
+        the bytes from an IOBytes in which a ZipFile wrote its contents, or the bytes
+        read directly from a zip file on disk
+    """
+    b64 = base64.b64encode(zip_bytes).decode()
+    href = f"<a href=\"data:file/zip;base64,{b64}\" download='{filename}'>\
+        Click to download {filename}\
+    </a>"
+    st.sidebar.markdown(href, unsafe_allow_html=True)
+
+
+def link_to_zip_download(zip_to_download: pathlib.Path):
+    """Generates a link allowing the data in a given zip to be downloaded
+
+    Parameters
+    ----------
+    zip_to_download:  pathlib.Path
+        path to zip
+
+    """
+    filename = os.path.basename(zip_to_download)
+    with open(zip_to_download, "rb") as f:
+        zip_bytes = f.read()
+    link_to_zipbuffer_download(filename, zip_bytes)
+
+
+def build_pseudonymised_file_name(ds_input: pydicom.dataset.Dataset):
+    """[summary] Provide full path to the intended destination of the pseudonymised file
+
+    Parameters
+    ----------
+    ds_input : pydicom.dataset.Dataset
+        the non-pseudonymised DICOM data
+
+    Returns
+    -------
+    str
+        full path to the individual pseudonymised file, which can be read later
+        for addition to a zip
+    """
+    pseudo_sop_instance_uid = pseudonymisation_api.pseudonymisation_dispatch["UI"](
+        ds_input.SOPInstanceUID
+    )
+
+    sop_class_uid: pydicom.dataelem.DataElement = ds_input.SOPClassUID
+
+    mode_prefix = DICOM_SOP_CLASS_NAMES_MODE_PREFIXES[
+        sop_class_uid.name  # pylint: disable = no-member
```

Comment:
It's interesting that pylint couldn't find this. I wonder if there's something we can do to help pylint here. Maybe a type hint might bump it in the right direction...